### DISCUSSION TITLE
Unexpected retry behavior

### DISCUSSION THREAD

**Discussion by paymog** [enisdenjo/graphql-ws#355] [Q&A] [answered]
**Screenshot**
<img width="555" alt="image" src="https://user-images.githubusercontent.com/1986950/164946914-fc3a9227-c4e5-4668-8d56-e2ac5ea3976e.png">

Client initialization:

```
const wsLink = new GraphQLWsLink(
  // https://stackoverflow.com/a/60958527/1489726
  process.browser
    ? createClient({
        url: settings?.graphql?.wsUri,
        retryAttempts: 10,
        retryWait: async (retries) => {
          console.log(`retry attempt #${retries}`);

          // taken directly from the docs at https://github.com/enisdenjo/graphql-ws#retry-strategy
          // after the server becomes ready, wait for a second + random 1-4s timeout
          // (avoid DDoSing yourself) and try connecting again
          await new Promise((resolve) => setTimeout(resolve, 1000 + Math.random() * 3000));
        },
      })
    : null,
);
```

**Expected Behaviour**
I expect multiple retry attempts to happen and multiple log lines of the form `retry attempt #...` to be printed in the browser

**Actual Behaviour**
It seems only 1 retry attempt is attempted because I only see 1 log line

**Debug Information**
1. start up apollo server
2. start up app with the above websocket client
3. see the webscoket is connected
4. stop apollo server
5. see that only 1 retry attempt is made

**Further Information**
Is it possible that a server that's down is considered a fatal error? If so, are there docs which explain that? It's not obvious if a non-responsive server is fatal based on the comments in `createClient`

https://github.com/enisdenjo/graphql-ws/blob/e9f7c94abd834b76bdd32a1e9a1d88ca2a73cedb/src/client.ts#L328-L337

**Comment by enisdenjo** [comment, marked as answer]
All non-close-event connection problems are considered fatal by default. Your client seems to be throwing `undefined`, which is not a [WebSocket CloseEvent](https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent).

If you want the client to retry regardless, you can implement your own [`shouldRetry` option](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/client.ClientOptions.md#shouldretry) and have it return `true` always (or with pre [v5.8.0](https://github.com/enisdenjo/graphql-ws/releases/tag/v5.8.0) the now [deprecated `isFatalConnectionProblem` option](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/client.ClientOptions.md#isfatalconnectionproblem) and return `false` always).
